src/adt/edgeset.c

```diff
@@ -871,3 +887,997 @@ edge_set_ordered_iter_next(struct edge_ordered_iter *eoi, struct fsm_edge *e)
+
+	while (state_set_next(&si, &state)) {
+		if (!edge_set_add(setp, alloc, symbol, state)) {
+			return 0;
+		}
+	}
+
+	return 1;
+}
+
+int
+edge_set_find(const struct edge_set *set, unsigned char symbol,
+	struct fsm_edge *e)
+{
+	const struct edge_group *eg;
+	size_t i;
+	if (set == NULL) {
+		return 0;
+	}
+
+#if LOG_BITSET
+	fprintf(stderr, " -- edge_set_find: symbol 0x%x on %p\n",
+	    symbol, (void *)set);
+#endif
+
+	for (i = 0; i < set->count; i++) {
+		eg = &set->groups[i];
+		if (SYMBOLS_GET(eg->symbols, symbol)) {
+			e->state = eg->to;
+			e->symbol = symbol;
+			return 1;
+		}
+	}
+
+	return 0;		/* not found */
+}
+
+int
+edge_set_contains(const struct edge_set *set, unsigned char symbol)
+{
+	const struct edge_group *eg;
+	size_t i;
+	if (edge_set_empty(set)) {
+		return 0;
+	}
+
+#if LOG_BITSET
+	fprintf(stderr, " -- edge_set_contains: symbol 0x%x on %p\n",
+	    symbol, (void *)set);
+#endif
+
+	for (i = 0; i < set->count; i++) {
+		eg = &set->groups[i];
+		if (SYMBOLS_GET(eg->symbols, symbol)) {
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+int
+edge_set_hasnondeterminism(const struct edge_set *set, struct bm *bm)
+{
+	size_t i, w_i;
+	assert(bm != NULL);
+
+#if LOG_BITSET
+	fprintf(stderr, " -- edge_set_hasnondeterminism: on %p\n",
+	    (void *)set);
+#endif
+
+	dump_edge_set(set);
+
+	if (edge_set_empty(set)) {
+		return 0;
+	}
+
+	for (i = 0; i < set->count; i++) {
+		const struct edge_group *eg = &set->groups[i];
+		for (w_i = 0; w_i < 4; w_i++) {
+			const uint64_t cur = eg->symbols[w_i];
+			size_t b_i;
+			if (cur == 0) {
+				continue;
+			}
+
+#if LOG_BITSET > 1
+			fprintf(stderr, " -- eshnd: [0x%lx, 0x%lx, 0x%lx, 0x%lx] + group %ld cur %ld: 0x%lx\n",
+			    eg->symbols[0], eg->symbols[1],
+			    eg->symbols[2], eg->symbols[3],
+			    i, w_i, cur);
+#endif
+
+			for (b_i = 0; b_i < 64; b_i++) {
+				if (cur & (1ULL << b_i)) {
+					const size_t bit = 64*w_i + b_i;
+					if (bm_get(bm, bit)) {
+						return 1;
+					}
+					bm_set(bm, bit);
+				}
+			}
+		}
+	}
+
+	return 0;
+}
+
+int
+edge_set_transition(const struct edge_set *set, unsigned char symbol,
+	fsm_state_t *state)
+{
+	/*
+	 * This function is meaningful for DFA only; we require a DFA
+	 * by contract in order to identify a single destination state
+	 * for a given symbol.
+	 */
+	struct fsm_edge e;
+	if (!edge_set_find(set, symbol, &e)) {
+		return 0;
+	}
+	*state = e.state;
+	return 1;
+}
+
+size_t
+edge_set_count(const struct edge_set *set)
+{
+	size_t res = 0;
+	size_t i;
+
+#if LOG_BITSET
+	fprintf(stderr, " -- edge_set_count: on %p\n",
+	    (void *)set);
+#endif
+
+	/* This does not call edge_set_empty directly
+	 * here, because that does the same scan as below,
+	 * it just exits at the first label found. */
+	if (set == NULL) {
+		return 0;
+	}
+
+	for (i = 0; i < set->count; i++) {
+		size_t w_i;
+		const struct edge_group *eg = &set->groups[i];
+		for (w_i = 0; w_i < 4; w_i++) {
+			/* TODO: res += popcount64(w) */
```

Comment:
We could use a compiler intrinsic/builtin for 64-bit popcount here instead.

Reply:
Nice.